crates/matrix-sdk/src/oidc.rs

```diff
@@ -0,0 +1,966 @@
+    pub async fn url_for_scope_with_authorization_code(
+        &self,
+        scope: &Scope,
+        redirect_uri: &Url,
+    ) -> Result<OidcAuthorizationData, OidcError> {
+        let provider_metadata = self.provider_metadata().await?;
+        let client_credentials = self.client_credentials().ok_or(OidcError::NotAuthenticated)?;
+
+        let authorization_data = AuthorizationRequestData {
+            client_id: client_credentials.client_id(),
+            code_challenge_methods_supported: provider_metadata
+                .code_challenge_methods_supported
+                .as_deref(),
+            scope,
+            redirect_uri,
+            prompt: None,
+        };
+
+        // TODO: use PAR if possible.
+
+        let (url, validation_data) = build_authorization_url(
+            provider_metadata.authorization_endpoint().clone(),
+            authorization_data,
+            &mut rand::thread_rng(),
+        )?;
+
+        let state = validation_data.state.clone();
+
+        self.client.inner.oidc_validation_data.lock().await.insert(state.clone(), validation_data);
+
+        Ok(OidcAuthorizationData { url, state })
+    }
+
+    /// Finish authorizing a given scope by using the Authorization Code flow.
+    ///
+    /// This method should be called after the URL returned by
+    /// [`Oidc::url_for_scope_with_authorization_code()`] has been presented and
+    /// the user has been redirected to the redirect URI.
+    ///
+    /// # Arguments
+    ///
+    /// * `code` - The code received as part of the redirect URI when the
+    ///   authorization was successful.
+    ///
+    /// * `state` - The unique string received as part of the redirect URI when
+    ///   the authorization was successful
+    ///
+    /// [`ClientErrorCode`]: matrix_sdk::oidc::types::errors::ClientErrorCode
+    pub async fn finish_authorization_with_authorization_code(
+        &self,
+        auth_response: AuthorizationResponse,
+    ) -> Result<AccessTokenResponse, OidcError> {
+        let provider_metadata = self.provider_metadata().await?;
+        let client_credentials = self.client_credentials().ok_or(OidcError::NotAuthenticated)?;
+
+        let validation_data = self
+            .client
+            .inner
+            .oidc_validation_data
+            .lock()
+            .await
+            .remove(&auth_response.state)
+            .ok_or(OidcError::InvalidState)?;
+
+        let (response, id_token) = access_token_with_authorization_code(
+            &self.http_service(),
+            client_credentials.clone(),
+            provider_metadata.token_endpoint(),
+            auth_response.code,
+            validation_data,
+            None,
+            Utc::now(),
+            &mut rand::thread_rng(),
+        )
+        .await?;
+
+        let tokens = SessionTokens {
+            access_token: response.access_token.clone(),
+            refresh_token: response.refresh_token.clone(),
+        };
```

Comment:
It feels like this should be an enum instead, with a variant for 'classic' session and a variant for OIDC sessions with:

 - the actual scope (in case the server gave less scope than requested)
 - the expiration
 - the id_token
 - access + refresh token

Reply:
Storing of authentication data has been reworked in #2081 and this PR has been updated to use it.